### PR TITLE
Add configurable ackDeadline to GCP Pub/Sub component

### PR DESCRIPTION
**Pull request by mahparaashley**
# Description

(`ackDeadline`) for the GCP Pub/Sub component within Dapr's components-contrib repository. Previously, the acknowledgment deadline was hardcoded to 20 seconds. This change allows users to specify a custom acknowledgment deadline through the component's metadata, thus providing greater flexibility to tailor behavior to specific workload requirements.

# Motivation
In many scenarios, the hardcoded 20-second deadline may not be optimal. Applications with varying processing times may require longer deadlines to prevent premature message redelivery, which can lead to increased costs and workload on services. Making this value configurable allows users to optimize their systems based on actual performance and needs.

# Changes
Metadata Structure: Added an AckDeadline field to the metadata struct to hold the user-defined acknowledgment deadline.
Default Value: Set a default acknowledgment deadline (defaultAckDeadline) to 20 seconds to maintain backward compatibility.
Metadata Parsing: Updated the createMetadata function to parse the ackDeadline from the component's metadata, with validation to ensure it is a positive integer.
Subscription Configuration: Modified the subscription configuration logic in the ensureSubscription method to utilize the user-configured ackDeadline.
Testing: Enhanced the test suite to include tests for validating the behavior with a specified ackDeadline, handling of invalid values, and default behavior when not specified.

# Testing
Comprehensive tests were added to validate:

The correct parsing and application of a valid ackDeadline.
Proper error handling when an invalid ackDeadline is provided.
The use of the default ackDeadline when none is specified.
These tests ensure that the new functionality behaves as expected and that existing functionalities remain unaffected.

# Impact
This change is backward compatible. Existing configurations without a specified ackDeadline will use the default value, ensuring that current implementations will operate without modifications. New users can specify an ackDeadline to suit their specific needs.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_
https://github.com/dapr/components-contrib/issues/3421

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pubsub/gcp/pubsub/pubsub.go [hunk 1]
```diff
@@ -82,6 +83,10 @@ type WhatNow struct {
 	Type string `json:"type"`
 }
 
+type metadata struct {
+    AckDeadline int `mapstructure:"ackDeadline"`
```
Comment (ItalyPaleAle) [inline pubsub/gcp/pubsub/pubsub.go:87]: Please use type as `time.Duration` so Dapr can be smart and interpret it as seconds or Go durations

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Thanks for this PR!

Also, please fix the DCO: https://github.com/dapr/dapr/blob/master/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work

File: pubsub/gcp/pubsub/pubsub.go (hunk 1 above)
Comment (mahparaashley) [reply pubsub/gcp/pubsub/pubsub.go]: Thank you for your feedback @ItalyPaleAle, I've updated the type to `time.Duration`

**Comment by yaron2** [comment]
@mahparaashley please resolve the conflict

**Comment by berndverst** [comment]
This PR is in rough shape. Why was the `metadata` struct type moved into `pubsub.go` which is now a duplicate from the one declared in the same package in `metadata.go` ?

I'll try to fix it up. Please be more conscious of compiler / linter errors. You can run `make test` and `make lint` locally.

**Comment by mahparaashley** [comment]
> This PR is in rough shape. Why was the `metadata` struct type moved into `pubsub.go` which is now a duplicate from the one declared in the same package in `metadata.go` ?
> 
> I'll try to fix it up. Please be more conscious of compiler / linter errors. You can run `make test` and `make lint` locally.

[berndverst](https://github.com/berndverst) Thank you for your feedback. This was my first time working with Go, and I am still learning. I will take a look at it over the weekend.

**Comment by berndverst** [comment]
> > This PR is in rough shape. Why was the `metadata` struct type moved into `pubsub.go` which is now a duplicate from the one declared in the same package in `metadata.go` ?
> > I'll try to fix it up. Please be more conscious of compiler / linter errors. You can run `make test` and `make lint` locally.
> 
> [berndverst](https://github.com/berndverst) Thank you for your feedback. This was my first time working with Go, and I am still learning. I will take a look at it over the weekend.

It's ok I understand. We are now beyond code freeze. I pushed the fixes to your PR so I can merge the PR into the release. Otherwise you would have to wait 3 months for this feature!

**Comment by berndverst** [comment]
@mahparaashley for the changes I made primarily look at this commit. https://github.com/dapr/components-contrib/pull/3422/commits/8f214c8aa9a9bcf3d142d67c70a5909ec797fb08

Since you used `time.Duration` for the metadata the value does not need to be converted to metadata explicitly. The way our code works, this allows users to specify durations like "1m30s" or "5s" or "6h" etc. But we also support unit less numbers like "12" to simply refer to seconds. These duration strings are preferred - so you used the correct data type here. When you to value comparison after metadata parsing however you need to also convert to the relevant duration type.

Those were the main changes besides some incorrect instantiation of types or declarations of properties.

I pushed my changes to your PR and will merge it. This way you can get credit for the contribution, but unfortunately I cannot wait for you to fix it up at this point given our code freeze.

**Review by berndverst — APPROVED** [review]

**Comment by mahparaashley** [comment]
@berndverst Thank you so much!

**Comment by marcduiker** [comment]
@holopin-bot @mahparaashley Thank you!

**Pull request merged**